File: split/prefixedcommand.pas
procedure prefixedcommand;
label
  30, 10;
var
  a: smallnumber;
  f: internalfontnumber;
  j: halfword;
  k: fontindex;
  p, q: halfword;
  n: integer;
  e: boolean;
begin
  a := 0;
  while curcmd = 93 do
  begin
    if not odd(a div curchr) then
      a := a + curchr;
    {404:}
repeat
      getxtoken;
    until (curcmd <> 10) and (curcmd <> 0){:404}
;
    if curcmd <= 70 then{1212:}

    begin
      begin
        if interaction = 3 then;
        printnl(262);
        print(1178);
      end;
      printcmdchr(curcmd, curchr);
      printchar(39);
      begin
        helpptr := 1;
        helpline[0] := 1179;
      end;
      backerror;
      goto 10;
    end{:1212}
;
  end;
  {1213:}
if (curcmd <> 97) and (a mod 4 <> 0) then
  begin
    begin
      if interaction = 3 then;
      printnl(262);
      print(685);
    end;
    printesc(1170);
    print(1180);
    printesc(1171);
    print(1181);
    printcmdchr(curcmd, curchr);
    printchar(39);
    begin
      helpptr := 1;
      helpline[0] := 1182;
    end;
    error;
  end{:1213}
;
  {1214:}
if eqtb[5306].int <> 0 then
    if eqtb[5306].int < 0 then
    begin
      if (a >= 4) then
        a := a - 4;
    end
    else if not (a >= 4) then
      a := a + 4{:1214}
;
  case curcmd of
    {1217:}
87: if (a >= 4) then
        geqdefine(3934, 120, curchr)
      else
        eqdefine(3934, 120, curchr);
    {:1217}
{1218:}
97:
    begin
      if odd(curchr) and not (a >= 4) and (eqtb[5306].int >= 0) then
        a := a + 4;
      e := (curchr >= 2);
      getrtoken;
      p := curcs;
      q := scantoks(true, e);
      if (a >= 4) then
        geqdefine(p, 111 + (a mod 4), defref)
      else
        eqdefine(p, 111 + (a mod 4), defref);
    end;
    {:1218}
{1221:}
94:
    begin
      n := curchr;
      getrtoken;
      p := curcs;
      if n = 0 then
      begin
        repeat
          gettoken;
        until curcmd <> 10;
        if curtok = 3133 then
        begin
          gettoken;
          if curcmd = 10 then
            gettoken;
        end;
      end
      else
      begin
        gettoken;
        q := curtok;
        gettoken;
        backinput;
        curtok := q;
        backinput;
      end;
      if curcmd >= 111 then
        mem[curchr].hh.lh := mem[curchr].hh.lh + 1;
      if (a >= 4) then
        geqdefine(p, curcmd, curchr)
      else
        eqdefine(p, curcmd, curchr);
    end;
    {:1221}
{1224:}
95:
    begin
      n := curchr;
      getrtoken;
      p := curcs;
      if (a >= 4) then
        geqdefine(p, 0, 256)
      else
        eqdefine(p, 0, 256);
      scanoptionalequals;
      case n of
        0:
        begin
          scancharnum;
          if (a >= 4) then
            geqdefine(p, 68, curval)
          else
            eqdefine(p, 68, curval);
        end;
        1:
        begin
          scanfifteenbitint;
          if (a >= 4) then
            geqdefine(p, 69, curval)
          else
            eqdefine(p, 69, curval);
        end;
        others:
        begin
          scaneightbitint;
          case n of
            2: if (a >= 4) then
                geqdefine(p, 73, 5318 + curval)
              else
                eqdefine(p, 73,
                  5318 + curval);
            3: if (a >= 4) then
                geqdefine(p, 74, 5851 + curval)
              else
                eqdefine(p, 74, 5851 + curval);
            4: if (a >= 4) then
                geqdefine(p, 75, 2900 + curval)
              else
                eqdefine(p, 75, 2900 + curval);
            5: if (a >= 4) then
                geqdefine(p, 76, 3156 + curval)
              else
                eqdefine(p, 76, 3156 + curval);
            6: if (a >= 4) then
                geqdefine(p, 72, 3422 + curval)
              else
                eqdefine(p, 72, 3422 + curval);
          end;
        end
      end;
    end;
    {:1224}
{1225:}
96:
    begin
      scanint;
      n := curval;
      if not scankeyword(841) then
      begin
        begin
          if interaction = 3 then;
          printnl(262);
          print(1072);
        end;
        begin
          helpptr := 2;
          helpline[1] := 1199;
          helpline[0] := 1200;
        end;
        error;
      end;
      getrtoken;
      p := curcs;
      readtoks(n, p);
      if (a >= 4) then
        geqdefine(p, 111, curval)
      else
        eqdefine(p, 111, curval);
    end;
    {:1225}
{1226:}
71, 72:
    begin
      q := curcs;
      if curcmd = 71 then
      begin
        scaneightbitint;
        p := 3422 + curval;
      end
      else
        p := curchr;
      scanoptionalequals;
      {404:}
repeat
        getxtoken;
      until (curcmd <> 10) and (curcmd <> 0){:404}
;
      if curcmd <> 1 then{1227:}

      begin
        if curcmd = 71 then
        begin
          scaneightbitint;
          curcmd := 72;
          curchr := 3422 + curval;
        end;
        if curcmd = 72 then
        begin
          q := eqtb[curchr].hh.rh;
          if q = 0 then
            if (a >= 4) then
              geqdefine(p, 101, 0)
            else
              eqdefine(p, 101, 0)
          else
          begin
            mem[q].hh.lh := mem[q].hh.lh + 1;
            if (a >= 4) then
              geqdefine(p, 111, q)
            else
              eqdefine(p, 111, q);
          end;
          goto 30;
        end;
      end{:1227}
;
      backinput;
      curcs := q;
      q := scantoks(false, false);
      if mem[defref].hh.rh = 0 then
      begin
        if (a >= 4) then
          geqdefine(p, 101, 0)
        else
          eqdefine(p, 101, 0);
        begin
          mem[defref].hh.rh := avail;
          avail := defref;
          {dynused:=dynused-1;}
end;
      end
      else
      begin
        if p = 3413 then
        begin
          mem[q].hh.rh := getavail;
          q := mem[q].hh.rh;
          mem[q].hh.lh := 637;
          q := getavail;
          mem[q].hh.lh := 379;
          mem[q].hh.rh := mem[defref].hh.rh;
          mem[defref].hh.rh := q;
        end;
        if (a >= 4) then
          geqdefine(p, 111, defref)
        else
          eqdefine(p, 111, defref);
      end;
    end;
    {:1226}
{1228:}
73:
    begin
      p := curchr;
      scanoptionalequals;
      scanint;
      if (a >= 4) then
        geqworddefine(p, curval)
      else
        eqworddefine(p, curval);
    end;
    74:
    begin
      p := curchr;
      scanoptionalequals;
      scandimen(false, false, false);
      if (a >= 4) then
        geqworddefine(p, curval)
      else
        eqworddefine(p, curval);
    end;
    75, 76:
    begin
      p := curchr;
      n := curcmd;
      scanoptionalequals;
      if n = 76 then
        scanglue(3)
      else
        scanglue(2);
      trapzeroglue;
      if (a >= 4) then
        geqdefine(p, 117, curval)
      else
        eqdefine(p, 117, curval);
    end;
    {:1228}
{1232:}
85:
    begin
      {1233:}
if curchr = 3983 then
        n := 15
      else if curchr = 5007 then
        n := 32768
      else if curchr = 4751 then
        n := 32767
      else if curchr = 5574 then
        n := 16777215
      else
        n := 255{:1233}
;
      p := curchr;
      scancharnum;
      p := p + curval;
      scanoptionalequals;
      scanint;
      if ((curval < 0) and (p < 5574)) or (curval > n) then
      begin
        begin
          if interaction = 3 then;
          printnl(262);
          print(1201);
        end;
        printint(curval);
        if p < 5574 then
          print(1202)
        else
          print(1203);
        printint(n);
        begin
          helpptr := 1;
          helpline[0] := 1204;
        end;
        error;
        curval := 0;
      end;
      if p < 5007 then
        if (a >= 4) then
          geqdefine(p, 120, curval)
        else
          eqdefine(p, 120,
            curval)
      else if p < 5574 then
        if (a >= 4) then
          geqdefine(p, 120, curval + 0)
        else
          eqdefine(p, 120, curval + 0)
      else if (a >= 4) then
        geqworddefine(p, curval)
      else
        eqworddefine(p, curval);
    end;
    {:1232}
{1234:}
86:
    begin
      p := curchr;
      scanfourbitint;
      p := p + curval;
      scanoptionalequals;
      scanfontident;
      if (a >= 4) then
        geqdefine(p, 120, curval)
      else
        eqdefine(p, 120, curval);
    end;
    {:1234}
{1235:}
89, 90, 91, 92: doregistercommand(a);
    {:1235}
{1241:}
98:
    begin
      scaneightbitint;
      if (a >= 4) then
        n := 256 + curval
      else
        n := curval;
      scanoptionalequals;
      if setboxallowed then
        scanbox(1073741824 + n)
      else
      begin
        begin
          if interaction = 3 then;
          printnl(262);
          print(680);
        end;
        printesc(536);
        begin
          helpptr := 2;
          helpline[1] := 1210;
          helpline[0] := 1211;
        end;
        error;
      end;
    end;
    {:1241}
{1242:}
79: alteraux;
    80: alterprevgraf;
    81: alterpagesofar;
    82: alterinteger;
    83: alterboxdimen;
    {:1242}
{1248:}
84:
    begin
      scanoptionalequals;
      scanint;
      n := curval;
      if n <= 0 then
        p := 0
      else
      begin
        p := getnode(2 * n + 1);
        mem[p].hh.lh := n;
        for j := 1 to n do
        begin
          scandimen(false, false, false);
          mem[p + 2 * j - 1].int := curval;
          scandimen(false, false, false);
          mem[p + 2 * j].int := curval;
        end;
      end;
      if (a >= 4) then
        geqdefine(3412, 118, p)
      else
        eqdefine(3412, 118, p);
    end;
    {:1248}
{1252:}
99: if curchr = 1 then
      begin
        newpatterns;
        goto 30;
        begin
          if interaction = 3 then;
          printnl(262);
          print(1215);
        end;
        helpptr := 0;
        error;
        repeat
          gettoken;
        until curcmd = 2;
        goto 10;
      end
      else
      begin
        newhyphexceptions;
        goto 30;
      end;
    {:1252}
{1253:}
77:
    begin
      findfontdimen(true);
      k := curval;
      scanoptionalequals;
      scandimen(false, false, false);
      fontinfo[k].int := curval;
    end;
    78:
    begin
      n := curchr;
      scanfontident;
      f := curval;
      scanoptionalequals;
      scanint;
      if n = 0 then
        hyphenchar[f] := curval
      else
        skewchar[f] := curval;
    end;
    {:1253}
{1256:}
88: newfont(a);
    {:1256}
{1264:}
100: newinteraction;
    {:1264}
others: confusion(1177)
  end;
  30:
    {1269:}
if aftertoken <> 0 then
    begin
      curtok := aftertoken;
      backinput;
      aftertoken := 0;
    end{:1269}
;
  10: ;
end;

File: cpp/prefixedcommand.cpp
#include "prefixedcommand.h"
#include "getxtoken.h"
#include "printnl.h"
#include "print.h"
#include "printcmdchr.h"
#include "printchar.h"
#include "backerror.h"
#include "printesc.h"
#include "error.h"
#include "getrtoken.h"
#include "geqdefine.h"
#include "scantoks.h"
#include "gettoken.h"
#include "eqdefine.h"
#include "scanoptionalequals.h"
#include "backinput.h"
#include "scancharnum.h"
#include "scanfifteenbitint.h"
#include "scaneightbitint.h"
#include "scanint.h"
#include "readtoks.h"
#include "scankeyword.h"
#include "getavail.h"
#include "eqworddefine.h"
#include "scandimen.h"
#include "scanglue.h"
#include "trapzeroglue.h"
#include "geqworddefine.h"
#include "scancharnum.h"
#include "scanfourbitint.h"
#include "scanfontident.h"
#include "printint.h"
#include "doregistercommand.h"
#include "scanbox.h"
#include "alteraux.h"
#include "alterprevgraf.h"
#include "alterpagesofar.h"
#include "alterboxdimen.h"
#include "alterinteger.h"
#include "getnode.h"
#include "newpatterns.h"
#include "newhyphexceptions.h"
#include "findfontdimen.h"
#include "newfont.h"
#include "newinteraction.h"
#include "confusion.h"
#include "texte.h"

void prefixedcommand(void)
{
	smallnumber a;
	internalfontnumber f;
	halfword j;
	fontindex k;
	halfword p, q;
	int n;
	bool e;
	a = 0;
	while (curcmd == prefix)
	{
		if ((a/curchr)%2 == 0)
			a += curchr;
		do
			getxtoken();
		while (curcmd == spacer || curcmd == escape);
		if (curcmd <= max_non_prefixed_command)
		{
			printnl("! ");
			print("You can't use a prefix with `");
			printcmdchr(curcmd, curchr);
			printchar('\''); 
			helpptr = 1;
			helpline[0] = "I'll pretend you didn't say \\long or \\outer or \\global.";
			backerror;
			return;
		}
	}
	if (curcmd != def && a%4)
	{
		printnl("! ");
		print("You can't use `");
		printesc("long");
		print("' or `");
		printesc("outer");
		print("' with `");
		printcmdchr(curcmd, curchr);
		printchar('\'');
		helpptr = 1;
		helpline[0] = "I'll pretend you didn't say \\long or \\outer here.";
		error();
	}
	if (int_par(global_defs_code))
		if (int_par(global_defs_code) < 0)
		{
			if (a >= 4)
				a -= 4;
		}
		else 
			if (a < 4)
			a += 4;
	switch (curcmd)
	{
		case set_font: 
			if (a >= 4)
				geqdefine(cur_font_loc, data, curchr);
			else
				eqdefine(cur_font_loc, data, curchr);
			break;
		case def:
			if (curchr%2 & a < 4 && int_par(global_defs_code) >= 0)
				a += 4;
			e = curchr >= 2;
			getrtoken();
			p = curcs;
			q = scantoks(true, e);
			if (a >= 4)
				geqdefine(p, 111+a%4, defref); // a%4 = 0:call 1:long_call 2:outer_call 3:long_outer_call
			else
				eqdefine(p, 111+a%4, defref); // a%4 = 0:call 1:long_call 2:outer_call 3:long_outer_call
			break;
		case let:
			n = curchr;
			getrtoken();
			p = curcs;
			if (n == 0)
			{
				do
					gettoken();
				while (curcmd == spacer);
				if (curtok == 3133) // other_char + '='
				{
					gettoken();
					if (curcmd == spacer)
						gettoken();
				}
			}
			else
			{
				gettoken();
				q = curtok;
				gettoken();
				backinput();
				curtok = q;
				backinput;
			}
			if (curcmd >= call)
				info(curchr)++;
			if (a >= 4)
				geqdefine(p, curcmd, curchr);
			else
				eqdefine(p, curcmd, curchr);
			break;
		case shorthand_def:
			n = curchr;
			getrtoken();
			p = curcs;
			if (a >= 4)
				geqdefine(p, relax, 256);
			else
				eqdefine(p, relax, 256);
			scanoptionalequals();
			switch (n)
			{
				case 0:
					scancharnum();
					if (a >= 4)
						geqdefine(p, char_given, curval);
					else
						eqdefine(p, char_given, curval);
					break;
				case 1:
					scanfifteenbitint();
					if (a >= 4)
						geqdefine(p, math_given, curval);
					else
						eqdefine(p, math_given, curval);
					break;
				default:
					scaneightbitint();
					switch (n)
					{
						case 2: 
							if (a >= 4)
								geqdefine(p, assign_int, count_base+curval);
							else
								eqdefine(p, assign_int, count_base+curval);
							break;
						case 3: 
							if (a >= 4)
								geqdefine(p, assign_dimen, scaled_base+curval);
							else
								eqdefine(p, assign_dimen, scaled_base+curval);
							break;
						case 4: 
							if (a >= 4)
								geqdefine(p, assign_glue, skip_base+curval);
							else
								eqdefine(p, assign_glue, skip_base+curval);
							break;
						case 5: 
							if (a >= 4)
								geqdefine(p, assign_mu_glue, mu_skip_base+curval);
							else
								eqdefine(p, assign_mu_glue, mu_skip_base+curval);
							break;
						case 6: 
							if (a >= 4)
								geqdefine(p, assign_toks, toks_base+curval);
							else
								eqdefine(p, assign_toks, toks_base+curval);
					}
			}
			break;
		case read_to_cs:
			scanint();
			n = curval;
			if (!scankeyword("to")) 
			{
				printnl("! ");
				print("Missing `to' inserted");
				helpptr = 2;
				helpline[1] = "You should have said `\\read<number> to \\cs'.";
				helpline[0] = "I'm going to look for the \\cs now.";
				error();
			}
			getrtoken();
			p = curcs;
			readtoks(n, p);
			if (a >= 4)
				geqdefine(p, call, curval);
			else
				eqdefine(p, call, curval);
			break;
		case toks_register:
		case assign_toks:
			q = curcs;
			if (curcmd == toks_register)
			{
				scaneightbitint();
				p = toks_base+curval;
			}
			else
				p = curchr;
			scanoptionalequals();
			do
				getxtoken();
			while (curcmd == spacer || curcmd == escape);
			if (curcmd != left_brace)
			{
				if (curcmd == toks_register)
				{
					scaneightbitint();
					curcmd = assign_toks;
					curchr = toks_base+curval;
				}
				if (curcmd == assign_toks)
				{
					q = equiv(curchr);
					if (q == 0)
						if (a >= 4)
							geqdefine(p, undefined_cs, 0);
						else
							eqdefine(p, undefined_cs, 0);
					else
					{
						info(q)++;
						if (a >= 4)
							geqdefine(p, call, q);
						else
							eqdefine(p, call, q);
					}
					break;
				}
			}
			backinput();
			curcs = q;
			q = scantoks(false, false);
			if (link(defref) == 0)
			{
				if (a >= 4)
					geqdefine(p, undefined_cs, 0);
				else
					eqdefine(p, undefined_cs, 0);
				link(defref) = avail;
				avail = defref;
			}
			else
			{
				if (p == 3413) //match +'U'/85 ?
				{
					link(q) = getavail();
					q = link(q);
					info(q) = right_brace*0x01'00+'}';
					q = getavail();
					info(q) = left_brace*0x01'00+'{';
					link(q) = link(defref);
					link(defref) = q;
				}
				if (a >= 4)
					geqdefine(p, call, defref);
				else
					eqdefine(p, call, defref);
			}
			break;
		case assign_int:
			p = curchr;
			scanoptionalequals();
			scanint();
			if (a >= 4)
				geqworddefine(p, curval);
			else
				eqworddefine(p, curval);
			break;
		case assign_dimen:
			p = curchr;
			scanoptionalequals();
			scandimen(false, false, false);
			if (a >= 4)
				geqworddefine(p, curval);
			else
				eqworddefine(p, curval);
			break;
		case assign_glue:
		case assign_mu_glue:
			p = curchr;
			n = curcmd;
			scanoptionalequals();
			if (n == 76)
				scanglue(3);
			else
				scanglue(2);
			trapzeroglue();
			if (a >= 4)
				geqdefine(p, glue_ref, curval);
			else
				eqdefine(p, glue_ref, curval);
			break;
		case def_code:
			if (curchr == cat_code_base)
				n = 0x0F;
			else 
				if (curchr == math_code_base)
					n = 0x80'00;
				else 
					if (curchr == sf_code_base)
						n = 0x7F'FF;
					else 
						if (curchr == del_code_base)
							n = 0xFF'FF'FF;
						else
						n = 0xFF;
			p = curchr;
			scancharnum();
			p += curval;
			scanoptionalequals();
			scanint();
			if ((curval < 0 && p < del_code_base) || curval > n)
			{
				printnl("! ");
				print("Invalid code (");
				printint(curval);
				if (p < del_code_base)
					print("), should be in the range 0..");
				else
					print("//), should be at most ");
				printint(n);
				helpptr = 1;
				helpline[0] = "I'm going to use 0 instead of that illegal code value.";
				error;
				curval = 0;
			}
			if (p < math_code_base)
				if (a >= 4)
					geqdefine(p, data, curval);
				else
					eqdefine(p, data, curval);
			else 
				if (p < del_code_base)
					if (a >= 4)
						geqdefine(p, data, curval);
					else
						eqdefine(p, data, curval);
				else 
					if (a >= 4)
						geqworddefine(p, curval);
					else
						eqworddefine(p, curval);
			break;
		case def_family:
			p = curchr;
			scanfourbitint();
			p += curval;
			scanoptionalequals();
			scanfontident();
			if (a >= 4)
				geqdefine(p, data, curval);
			else
				eqdefine(p, data, curval);
			break;
		case register_:
		case advance:
		case multiply:
		case divide: 
			doregistercommand(a);
			break;
		case set_box:
			scaneightbitint();
			if (a >= 4)
				n = 0x1'00+curval;
			else
				n = curval;
			scanoptionalequals();
			if (setboxallowed)
				scanbox(0x40'00'00'00+n);
			else
			{
				printnl("! ");
				print("Improper ");
				printesc("setbox");
				helpptr = 2;
				helpline[1] = "Sorry, \\setbox is not allowed after \\halign in a display,";
				helpline[0] = "or between \\accent and an accented character.";
				error();
			}
			break;
		case set_aux: 
			alteraux();
			break;
		case set_prev_graf: 
			alterprevgraf();
			break;
		case set_page_dimen: 
			alterpagesofar();
			break;
		case set_page_int: 
			alterinteger();
			break;
		case set_box_dimen: 
			alterboxdimen();
			break;
		case set_shape:
			scanoptionalequals();
			scanint();
			n = curval;
			if (n <= 0)
				p = 0;
			else
			{
				p = getnode(2*n+1);
				info(p) = n;
				for (j = 1; j <= n; j++)
				{
					scandimen(false, false, false);
					mem[p+2*j-1].int_ = curval;
					scandimen(false, false, false);
					mem[p+2*j].int_ = curval;
				}
			}
			if (a >= 4)
				geqdefine(par_shape_loc, shape_ref, p);
			else
				eqdefine(par_shape_loc, shape_ref, p);
			break;
		case hyph_data:
			if (curchr == 1)
				newpatterns();
			else
				newhyphexceptions();
			break;
		case assign_font_dimen:
			findfontdimen(true);
			k = curval;
			scanoptionalequals();
			scandimen(false, false, false);
			fontinfo[k].int_ = curval;
			break;
		case assign_font_int:
			n = curchr;
			scanfontident();
			f = curval;
			scanoptionalequals();
			scanint();
			if (n == 0)
				hyphenchar[f] = curval;
			else
				skewchar[f] = curval;
			break;
		case def_font: 
			newfont(a);
			break;
		case set_interaction: 
			newinteraction();
			break;
		default: 
			confusion("prefix");
	}
	if (aftertoken)
	{
		curtok = aftertoken;
		backinput();
		aftertoken = 0;
	}
}
